### PR TITLE
148: Refactor output path into Destination

### PR DESCRIPTION
**Pull request by DanielJette**
### What does this change accomplish?

Prerequisite for #148 

### How have you achieved it?

This PR introduces two core changes:

1. File manipulation is now abstracted through the new `Destination` interface. This allows Testify to implement various alternative file output locations. The existing `data/data` and `sdcard` output destinations have been ported to the new interface. This allows for future destinations to be easily supported in the future. This change is backwards compatible.
2. The Testify Gradle Plugin can now write to AndroidManifest placeholder values in the target application. This means that custom attributes defined in `build.gradle` are now visible to the Library code. So, advanced configuration options can now be set in the `build.gradle` file and used by any Library code. In this PR, I have used this feature to pass the `useSdCard` configuration option to the Library and also to pass the current Gradle module name.

📔 In addition, a small refactoring to `ErrorCause` was made to support the two new exceptions added

### Scope of Impact and Testing instructions

- `ScreenshotTest` and `Reporter` now use the `Destination` interface. This should not affect users of Testify, but does impact developers and those who wish to subclass Testify classes.
- The tests have been updated to reflect these changes.

### Notice

> **Warning**
> This change must keep `main` in a shippable state; **it may be shipped without further notice**.

### CODE REVIEW

File: Library/src/main/java/dev/testify/output/DataDirectoryDestination.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    context: Context,
+    fileName: String,
+    extension: String = PNG_EXTENSION,
+    val key: String? = null,
+    val root: String? = null
+) : Destination {
+
+    companion object {
+        private const val DATA_DESTINATION_DIR = "images"
+        private const val LOG_TAG = "DataDirectory"
```
Comment (olegosipenko) [inline Library/src/main/java/dev/testify/output/DataDirectoryDestination.kt:23]: in the `SdCardDestination` this tag named `SdCardDestination` - maybe worth naming it here in the uniform way? 🤔

File: Library/src/main/java/dev/testify/output/Destination.kt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+     * Ensure the output destination directory exists, create if necessary
+     */
+    fun assureDestination(context: Context): Boolean
+}
+
+// TODO: class MediaStoreDestination : Destination
+
+private const val MANIFEST_DESTINATION_KEY = "dev.testify.destination"
```
[the remark this answers is not in the document]
Comment (DanielJette) [reply Library/src/main/java/dev/testify/output/Destination.kt:76]: Oops, that's a mistake. This should be used by ManifestHelpers.kt. I'll fix that

[the remark this answers is not in the document]
Comment (DanielJette) [reply Library/src/main/java/dev/testify/output/Destination.kt:76]: I would have named this `DataDirectoryDestination` but there's a lint warning for log tags over 23 characters, so I truncated it

File: Library/src/main/java/dev/testify/output/DataDirectoryDestination.kt (hunk 1 above)
Comment (DanielJette) [reply Library/src/main/java/dev/testify/output/DataDirectoryDestination.kt:23]: I would have named this `DataDirectoryDestination` but there's a lint warning for log tags over 23 characters, so I truncated it

**Review by olegosipenko — APPROVED** [review]

File: Ext/Fullscreen/src/main/java/dev/testify/capture/fullscreen/FullscreenCaptureMethod.kt [hunk 3]
```diff
@@ -85,22 +82,19 @@ fun fullscreenCapture(activity: Activity, targetView: View?): Bitmap {
...
+ * @throws Exception
  */
-private fun Context.assureScreenshotDirectory() {
-    if (assureScreenshotDirectory(this)) return
-
-    throw ScreenshotDirectoryNotFoundException(
-        useSdCard = useSdCard(InstrumentationRegistry.getArguments()),
-        path = getOutputDirectoryPath(this).absolutePath
-    )
+private fun Destination.assureScreenshotDirectory(context: Context) {
```
[the remark this answers is not in the document]
Comment (DanielJette) [reply Ext/Fullscreen/src/main/java/dev/testify/capture/fullscreen/FullscreenCaptureMethod.kt:95]: This is a convenience wrapper around `Destination.assureDestination()` which checks the return value of `assureDestination()` and throws an exception if it's false.

**Pull request merged**